File: legacy/packages/core/src/tools/pg-tx.ts
import { ActionRegistry, ToolDefinition, ActionContext } from "../types.js";
import { txHandler } from "../actions/tx/tx.js";

const txRegistry: ActionRegistry = {
    begin: txHandler,
    commit: txHandler,
    rollback: txHandler,
    savepoint: txHandler,
    release: txHandler,
    list: txHandler,
};

export const PgTxToolSchema = txHandler.schema;

export async function pgTxHandler(params: any, context: ActionContext) {
    const handler = txRegistry[params.action];
    if (!handler) {
        throw new Error(`Unknown action: ${params.action}`);
    }
    return await handler.handler(params, context);
}

export const pgTxTool: ToolDefinition = {
    name: "pg_tx",
    config: {
        description: `Transaction lifecycle management for multi-step database operations.

Actions:
  • begin: Start a new transaction, returns session_id for subsequent calls
  • commit: Commit all changes in the transaction (releases session)
  • rollback: Discard all changes in the transaction (releases session)
  • savepoint: Create a named savepoint within the transaction
  • release: Release (remove) a savepoint
  • list: Show all active sessions (useful for discovery/debugging)

Workflow:
  1. Call begin → receive session_id
  2. Use session_id in pg_query/pg_schema calls
  3. Call commit or rollback with session_id

Session Lifecycle:
  • TTL: Sessions auto-rollback after 30 minutes of inactivity
  • Limit: Maximum 10 concurrent sessions (prevents resource exhaustion)
  • Cleanup: Connections are destroyed on close (no state leakage)

Examples:
  {"action": "list"}
  {"action": "begin"}
  {"action": "begin", "options": {"isolation_level": "serializable"}}
  {"action": "commit", "session_id": "<id>"}`,
        inputSchema: PgTxToolSchema,
        destructiveHint: true, // Transaction control affects data state
    },
    handler: (context) => (params) => pgTxHandler(params, context),
};

File: coldquery/tools/pg_tx.py
from typing import Literal
from fastmcp.exceptions import ToolError
from coldquery.dependencies import CurrentActionContext
from coldquery.core.context import ActionContext
from coldquery.app import mcp
from coldquery.actions.tx.lifecycle import (
    begin_handler,
    commit_handler,
    rollback_handler,
    savepoint_handler,
    release_handler,
    list_handler,
)

TX_ACTIONS = {
    "begin": begin_handler,
    "commit": commit_handler,
    "rollback": rollback_handler,
    "savepoint": savepoint_handler,
    "release": release_handler,
    "list": list_handler,
}

@mcp.tool()
async def pg_tx(
    action: Literal["begin", "commit", "rollback", "savepoint", "release", "list"],
    session_id: str | None = None,
    isolation_level: str | None = None,
    savepoint_name: str | None = None,
    context: ActionContext = CurrentActionContext(),
) -> str:
    """Manage PostgreSQL transaction lifecycle.

    Actions:
    - begin: Start a new transaction, returns session_id
    - commit: Commit transaction and close session
    - rollback: Rollback transaction and close session
    - savepoint: Create a savepoint within a transaction
    - release: Release a savepoint
    - list: List all active sessions with metadata
    """
    handler = TX_ACTIONS.get(action)
    if not handler:
        raise ToolError(f"Unknown action: {action}")

    params = {
        "session_id": session_id,
        "isolation_level": isolation_level,
        "savepoint_name": savepoint_name,
    }

    return await handler(params, context)
